src/backend/distributed/executor/executor.c

```diff
@@ -1267,6 +1312,9 @@ ConnectionStateMachine(WorkerSession *session)
 
 			case MULTI_CONNECTION_CONNECTED:
 			{
+				/* if we're expanding the nodes in a transaction, use 2PC */
+				Use2PCIfParticipatingNodesExpanded(session);
```

Comment:
Should this be here, or later when we start using the connection?

Reply:
As far as I can see, it doesn't make any difference.

If we've opened a connection which expands the participating nodes, even if we don't use this specific connection, we have to activate 2PC because the command/execution needs that anyway. 

Since `Use2PCIfParticipatingNodesExpanded ` (or anything equivalent) is idempotent, it seems OK to have it here (because it seems easier to read code in this way instead of moving it to `TransactionStateMachine`?).

Reply:
Btw, I pushed one more commit, which moved this logic into `TransactionStateMachine`. I think we can revert if you don't like this way.